pcapgo/ngread.go

```diff
@@ -0,0 +1,599 @@
+			r.bigEndian = false
+		} else {
+			err = errors.New("Wrong byte order value in Section Header")
+			return
+		}
+		// Set length to remaining length (length - (type + lengthfield = 8) - 4 for byteOrderMagic)
+		r.currentBlock.length = r.getUint32(r.buf[4:8]) - 8 - 4
+		return
+	}
+	// Set length to remaining length (length - (type + lengthfield = 8)
+	r.currentBlock.length = r.getUint32(r.buf[4:8]) - 8
+	return
+}
+
+// readOption reads a single arbitrary option (type and value). If there is no space left for options and end of options is missing, it is faked.
+func (r *NgReader) readOption() (err error) {
+	if r.currentBlock.length == 4 {
+		// no more options
+		r.currentOption.code = ngOptionCodeEndOfOptions
+		return
+	}
+	if err = r.readBytes(r.buf[:4]); err != nil {
+		return err
+	}
+	r.currentBlock.length -= 4
+	r.currentOption.code = ngOptionCode(r.getUint16(r.buf[:2]))
+	length := r.getUint16(r.buf[2:4])
+	if r.currentOption.code == ngOptionCodeEndOfOptions {
+		if length != 0 {
+			return errors.New("End of Options must be zero length")
+		}
+		return
+	}
+	if length != 0 {
+		if length < uint16(cap(r.currentOption.value)) {
+			r.currentOption.value = r.currentOption.value[:length]
+		} else {
+			r.currentOption.value = make([]byte, length)
+		}
+		if err = r.readBytes(r.currentOption.value); err != nil {
+			return
+		}
+		//consume padding
+		padding := length % 4
+		if padding > 0 {
+			padding = 4 - padding
+			if _, err = r.r.Discard(int(padding)); err != nil {
+				return
+			}
+		}
+		r.currentBlock.length -= uint32(length + padding)
+	}
+	return
+}
+
+// readSectionHeader parses the full section header and implements section skipping in case of version mismatch
+// if needed, the first interface is read
+func (r *NgReader) readSectionHeader() (err error) {
+	if r.options.SectionEndCallback != nil && r.activeSection {
+		interfaces := make([]NgInterface, len(r.ifaces))
+		for i := range r.ifaces {
+			interfaces[i] = r.ifaces[i]
+		}
+		r.options.SectionEndCallback(interfaces, r.sectionInfo)
+	}
+	// clear the interfaces
+	r.ifaces = r.ifaces[:0]
+	r.activeSection = false
+
+RESTART:
+	// read major, minor, section length
+	if err = r.readBytes(r.buf[:12]); err != nil {
+		return
+	}
+	r.currentBlock.length -= 12
+
+	vMajor := r.getUint16(r.buf[0:2])
+	vMinor := r.getUint16(r.buf[2:4])
+	if vMajor != ngVersionMajor || vMinor != ngVersionMinor {
+		if !r.options.SkipUnknownVersion {
+			// Well the standard actually says to skip unknown version section headers,
+			// but this would mean user would be kept in the dark about whats going on...
+			return ErrNgVersionMismatch
+		}
+		if _, err = r.r.Discard(int(r.currentBlock.length)); err != nil {
+			return
+		}
+		if err = r.skipSection(); err != nil {
+			return
+		}
+		goto RESTART
+	}
+
+	var section NgSectionInfo
+
+	for {
+		if err = r.readOption(); err != nil {
+			return
+		}
+		switch r.currentOption.code {
+		case ngOptionCodeEndOfOptions:
+			goto DONE
```

Comment:
instead of `goto DONE` here, how about naming the for loop and doing a `break loop` instead?

Reply:
Well I found it this way clearer since the label is where the code is going to instead of (in the source code - not the generated code) jumping up to the loop header and then jumping to the end of the loop. But since it's all the same to me I'll change it.